### ISSUE TITLE
How can I check on the server whether a request is made by Turbo? aka „What happened to Turbolinks-Referrer?“

### ISSUE THREAD

**Issue by danieldiekmeier** [hotwired/turbo#195]
Hello! I'm in the process of upgrading an app from Turbolinks to Turbo, and I wondered that happened to the `Turbolinks-Referrer` header that was available before.

I'm using it as for a few optimizations, for example I'm sending my SVG sprite inline, but only on the first (non-Turbolinks) request. I'm using [Flask](https://flask.palletsprojects.com/en/1.1.x/), but the way this works is pretty straightforward:

```python
@app.before_request
def before_request():
    g.is_turbolinks = "Turbolinks-Referrer" in request.headers
```

and then, in my HTML templates, I can do:

```html
<div style="display: none;" data-turbolinks-permanent id="global-svg">
	{% if g.is_turbolinks %}
		<!-- no svg because it's a turbolinks response -->
	{% else %}
		{{ insert_inline("sprite/sprite.svg") }}
	{% endif %}
</div>
```

I noticed that Turbo Drive does currently not seem to send any special headers. I added this workaround in my JavaScript:

```js
document.addEventListener('turbo:before-fetch-request', (event) => {
  event.detail.fetchOptions.headers['Turbo-Referrer'] = window.location.href
})
```

I was just wondering if this was omitted on purpose, or if you didn't get to it yet. I think it was really useful and I'd love to see it make a comeback! Thanks.

**Comment by vwong** [comment]
It sends `Accept: text/vnd.turbo-stream.html`, but that's probably not what you want.

**Comment by danieldiekmeier** [comment]
@vwong I don't think that's correct. It only adds `text/vnd.turbo-stream.html` to the `Accept` header in some cases, for example when submitting a form element: https://turbo.hotwire.dev/handbook/streams#streaming-from-http-responses

For a standard Turbo Drive visit (just clicking a `<a>` on my site), it definitely doesn't send it:

<img width="274" alt="image" src="https://user-images.githubusercontent.com/2098462/110450086-2a349800-80c3-11eb-877c-1b66c1f49104.png">

But that's what I'm asking about.

**Comment by vwong** [comment]
Ah, That must've been fixed in a later version, I was using beta.3 which behaved differently.

**Comment by danieldiekmeier** [comment]
Ah yes, that seems possible, there is a mention of the change in the release notes! Thanks for clarifying. :)

**Comment by coorasse** [comment]
You have also `request.headers['Turbo-Frame']`

**Comment by danieldiekmeier** [comment]
@coorasse No, I don‘t! This is not about a Turbo Frame, this is about navigating with Turbo Drive, like I used to navigate with Turbolinks.

**Comment by JonathanLorimer** [comment]
I would also be interested in this. I asked a similar question on [discourse](https://discuss.hotwire.dev/t/how-can-i-tell-turbo-drive-is-working/2725). I don't think drive is working for me, and its hard to tell whether it is or not...

**Comment by danieldiekmeier** [comment]
@JonathanLorimer I don't think that your question is similar. You seem to have problems with getting Turbo working, at least consistently. But for me, Turbo does definitely work. My issue is a very specific question: It is missing one header that was/is present in Turbolinks.

**Comment by JonathanLorimer** [comment]
@danieldiekmeier ahhh, I misunderstood your problem, sorry for adding noise to the issue.

**Comment by kendagriff** [comment]
We're also interested in some way for the server to distinguish between a full page refresh and a request initiated by Turbo-Drive. Seems like adding a `Turbo-Drive` request header, in the spirit of the `Turbo-Frame` header would be helpful.

Any interest in this?

**Comment by zavan** [comment]
It looks like Turbo now adds a "X-Turbo-Request-Id" header: https://github.com/hotwired/turbo/blob/9e057f284a60a9597446b258d562cebd55be71d7/src/http/fetch.js#L12

**Comment by adrienpoly** [comment]
> 
>   def turbo_request?
>     request.headers.key?("X-Turbo-Request-Id")
>   end
> Works for my purposes.


if you use `turbo-rails` there is already a helper method `turbo_frame_request?` for that 

https://github.com/hotwired/turbo-rails/blob/0ea92f344395234dd7fc39b0ef9ab02388e33235/app/controllers/turbo/frames/frame_request.rb#L20-L39

**Comment by zavan** [comment]
> > 
> >   def turbo_request?
> >     request.headers.key?("X-Turbo-Request-Id")
> >   end
> > Works for my purposes.
> 
> 
> if you use `turbo-rails` there is already a helper method `turbo_frame_request?` for that 
> 
> https://github.com/hotwired/turbo-rails/blob/0ea92f344395234dd7fc39b0ef9ab02388e33235/app/controllers/turbo/frames/frame_request.rb#L20-L39
> 
> 
> 

Thanks, but I think this only works for turbo frames, I need to detect turbo stream refresh requests.

**Comment by elalemanyo** [comment]
@zavan Did you finally find a good solution? I had a similar situation and so far this is working quite well for me:

```ruby
def turbo_stream?
  request.format.turbo_stream? && response.media_type == Mime[:turbo_stream]
end
```

I don't know if it makes much sense but it works for me so far 😅
